### ISSUE TITLE
radtest returns "No 'known good'password found"

### ISSUE THREAD

**Issue by vladzr** [hacor/unifi-freeradius-ldap#12]
Hi. I've been using Google LDAP as backend auth mechanism for other services and when I tried this, I am unable to do successful authentication. I have verified the correctness of several Google accounts I've been entering, but when I do radtest or equivalent RADIUS utilities from my WiFi controller, I get errors. 

An ldapsearch returns complete info about the user's dn.
> LDAPTLS_CERT=ldap.crt LDAPTLS_KEY=ldap.key ldapsearch -H ldaps://ldap.google.com:636 -b dc=my,dc=org '(uid=user.name)'

Here's the output of my RADIUS queries.

> (2) Received Access-Request Id 144 from B.B.B.B:56091 to A.A.A.A:1812 length 93
(2)   User-Name = "user.name@my.org"
(2)   User-Password = “password”
(2)   NAS-IP-Address = B.B.B.B
(2)   NAS-Port = 0
(2)   Message-Authenticator = 0x2400162a35a563608281ba2c8bf51847
(2) # Executing section authorize from file /etc/freeradius/sites-enabled/default
(2)   authorize {
(2)     policy filter_username {
(2)       if (&User-Name) {
(2)       if (&User-Name)  -> TRUE
(2)       if (&User-Name)  {
(2)         if (&User-Name =~ / /) {
(2)         if (&User-Name =~ / /)  -> FALSE
(2)         if (&User-Name =~ /@[^@]*@/ ) {
(2)         if (&User-Name =~ /@[^@]*@/ )  -> FALSE
(2)         if (&User-Name =~ /\.\./ ) {
(2)         if (&User-Name =~ /\.\./ )  -> FALSE
(2)         if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\.(.+)$/))  {
(2)         if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\.(.+)$/))   -> FALSE
(2)         if (&User-Name =~ /\.$/)  {
(2)         if (&User-Name =~ /\.$/)   -> FALSE
(2)         if (&User-Name =~ /@\./)  {
(2)         if (&User-Name =~ /@\./)   -> FALSE
(2)       } # if (&User-Name)  = notfound
(2)     } # policy filter_username = notfound
(2)     [preprocess] = ok
(2)     [digest] = noop
(2) suffix: Checking for suffix after "@"
(2) suffix: Looking up realm "my.org" for User-Name = "user.name@my.org"
(2) suffix: Found realm "my.org"
(2) suffix: Adding Stripped-User-Name = "user.name"
(2) suffix: Adding Realm = "my.org"
(2) suffix: Authentication realm is LOCAL
(2)     [suffix] = ok
(2) eap: No EAP-Message, not doing EAP
(2)     [eap] = noop
(2)     [files] = noop
rlm_ldap (ldap): Reserved connection (9)
(2) ldap: Performing search in "dc=my,dc=org" with filter "(uid=user.name)", scope "sub"
(2) ldap: Waiting for search result...
(2) ldap: Search returned no results
rlm_ldap (ldap): Released connection (9)
Need 7 more connections to reach 10 spares
rlm_ldap (ldap): Opening additional connection (11), 1 of 29 pending slots used
rlm_ldap (ldap): Connecting to ldaps://ldap.google.com:636
rlm_ldap (ldap): Waiting for bind result...
rlm_ldap (ldap): Bind successful
(2)     [ldap] = notfound
(2)     [expiration] = noop
(2)     [logintime] = noop
(2) pap: WARNING: No "known good" password found for the user.  Not setting Auth-Type
(2) pap: WARNING: Authentication will fail unless a "known good" password is available
(2)     [pap] = noop
(2)     if (User-Password) {
(2)     if (User-Password)  -> TRUE
(2)     if (User-Password)  {
(2)       update control {
(2)         Auth-Type := ldap
(2)       } # update control = noop
(2)     } # if (User-Password)  = noop
(2)   } # authorize = ok
(2) Found Auth-Type = ldap
(2) # Executing group from file /etc/freeradius/sites-enabled/default
(2)   authenticate {
rlm_ldap (ldap): Reserved connection (8)
(2) ldap: Login attempt by "user.name"
(2) ldap: EXPAND (uid=%{%{Stripped-User-Name}:-%{User-Name}})
(2) ldap:    --> (uid=user.name)
(2) ldap: Performing search in "dc=my,dc=org" with filter "(uid=user.name)", scope "sub"
(2) ldap: Waiting for search result...
(2) ldap: Search returned no results
rlm_ldap (ldap): Released connection (8)
Need 6 more connections to reach 10 spares
rlm_ldap (ldap): Opening additional connection (12), 1 of 28 pending slots used
rlm_ldap (ldap): Connecting to ldaps://ldap.google.com:636
rlm_ldap (ldap): Waiting for bind result...
rlm_ldap (ldap): Bind successful
(2)     [ldap] = notfound
(2)   } # authenticate = notfound
(2) Failed to authenticate the user
(2) Using Post-Auth-Type Reject
(2) # Executing group from file /etc/freeradius/sites-enabled/default
(2)   Post-Auth-Type REJECT {
(2) attr_filter.access_reject: EXPAND %{User-Name}
(2) attr_filter.access_reject:    --> user.name@my.org
(2) attr_filter.access_reject: Matched entry DEFAULT at line 11
(2)     [attr_filter.access_reject] = updated
(2)     [eap] = noop
(2)     policy remove_reply_message_if_eap {
(2)       if (&reply:EAP-Message && &reply:Reply-Message) {
(2)       if (&reply:EAP-Message && &reply:Reply-Message)  -> FALSE
(2)       else {
(2)         [noop] = noop
(2)       } # else = noop
(2)     } # policy remove_reply_message_if_eap = noop
(2)   } # Post-Auth-Type REJECT = updated
(2) Delaying response for 1.000000 seconds
(1) Cleaning up request packet ID 17 with timestamp +769
Waking up in 0.9 seconds.
(2) Sending delayed response
(2) Sent Access-Reject Id 144 from A.A.A.A:1812 to B.B.B.B:56091 length 20
Waking up in 3.9 seconds.

Hoping you could help and will greatly appreciate your response. Thank you. :)

**Comment by hacor** [comment]
Hello @vladzurbano 

From what I see from this part of your logs:
```
(2) ldap: Login attempt by "user.name"
(2) ldap: EXPAND (uid=%{%{Stripped-User-Name}:-%{User-Name}})
(2) ldap: --> (uid=user.name)
(2) ldap: Performing search in "dc=my,dc=org" with filter "(uid=user.name)", scope "sub"
(2) ldap: Waiting for search result...
(2) ldap: Search returned no results
```

Is that the user ID is successfully handed over to Google LDAP, but that leads to no results.
So I think your Radius part is doing OK. Are you using organisational units? I'm not sure but I can imagine this could cause problems. Can you also post the log files (reports) from you Google Admin side on the LDAP requests

Good luck!

Hacor

**Comment by vladzr** [comment]
Thank you for the prompt response Hacor. We do have organisational units in place, but the permissions I have for the LDAP bind client can verify user passwords across the whole domain. I'll explore what parameters might be causing this but here's the output of from Audit Log.

LDAP search with (uid=user.name)successful. | Search successful | my_radius | 285157000656 |   | Dec 11, 2020, 11:35:12 AM GMT+8 | A.A.A.A | userPassword userPassword ntPassword radiusControlAttribute radiusRequestAttribute radiusReplyAttribute
-- | -- | -- | -- | -- | -- | -- | --

Cheers,
Vlad

**Comment by hacor** [comment]
@vladzurbano 
Sad it couldn't help. Here's my configuration of the LDAP client in Google Admin (terms might differ because Admin is not in English :-) )
```
- Login credentials check: Whole domain
- Read user credentials: Whole domain
- Read Group info: Enabled
```

The logs seem OK. The search was successful in Google service as well... You should get another line in the table when the user logged in (LDAP-connection with user ... was successful)

Hopefully we get this worked out

Best!
H

**Comment by vladzr** [comment]
@hacor 
I'm actually thinking now that it has something to do with how Google processes it. With only `Login credentials check: Whole domain` in place, I was able to authenticate successfully on a few requests. However, logs are filled initially with the `pap: WARNING: No "known good" password for the user.` before getting a successful auth and Access-Accept. 

I then tried giving permission to those three parameters you mentioned, and the result is the same. I was also able to see the LDAP logs you mentioned which tags the user that was able to login through LDAP. Although after multiple times of modifying the permissions, queries are getting rejected again. 

I might as well check the latency which might be dropping the connections prematurely. Thanks again!
